### ISSUE TITLE
Can You Log Select Statements?

### ISSUE THREAD

**Issue by chobo2** [thepirat000/Audit.NET#512]
Hi

I am wondering can this also log select statements for audit tracking?

**Comment by thepirat000** [comment]
I guess you are referring to Audit.EntityFramework.Core, if that's the case, you can use the low-level interceptor

https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.EntityFramework/README.md#low-level-command-interception

**Comment by chobo2** [comment]
> I guess you are referring to Audit.EntityFramework.Core, if that's the case, you can use the low-level interceptor
> 
> https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.EntityFramework/README.md#low-level-command-interception

Yes, my plan is too hookup this auditor with EF and SQL database, I just need to log everything including all select statements, inserts, deletes and updates.

is there more an example on how to do it with this low level command interpreter? It does not really show enough for me to understand right now.

Also can these all be saved in the same table that is generated? Looking to have more 1 audit table for everything.

**Comment by thepirat000** [comment]
Take into account that you cannot use the EntityFramework data provider to store the low-level interceptor events, since the EF data provider is exclusively for the high-level SaveChanges interception events.

So if you want to store low-level events to a SQL table, you can use the Audit.NET.SqlServer data provider.

For example, you can configure your DbContext to include both high-level and low-level as follows:

```c#
// Inheriting from AuditDbContext will intercept SaveChanges for high-level events. 
// Inherit from DbContext if you only want low-level events.
public class MyAuditedContext : AuditDbContext
{
    public DbSet<User> Users { get; set; }

    public DbSet<AuditEntity> Audits { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        if (!optionsBuilder.IsConfigured)
        {
            // Add the low-level interceptor
            optionsBuilder.AddInterceptors(new AuditCommandInterceptor());
            optionsBuilder.UseSqlServer(Settings.SqlConnectionString);
        }
    }
}

// Sample audit entity class to store in the database
[AuditIgnore]
public class AuditEntity
{
    [Key] public int Id { get; set; }
    public DateTime? DateTime { get; set; }
    public string Event { get; set; }
}
```

So you can configure the SQL data provider like this:

```c#
Audit.Core.Configuration.Setup()
    .UseSqlServer(cfg => cfg
        .ConnectionString(Settings.SqlConnectionString)
        .TableName("Audits")
        .IdColumnName("Id")
        .CustomColumn("DateTime", ev => DateTime.Now)
        .JsonColumnName("Event"));
```

**Comment by chobo2** [comment]
> ```cs
> Settings
> ```

Okay I will try that but where is Settings coming from?

Also where do you put the last chunk of code .Setup()?

My project is split into many different projects where the ApplicationDBContext is on 1 project and the Entities and the setup of the context is in a seperate project

looks something like this

```
public static IServiceCollection Add(this IServiceCollection, services, IConfigruation config)
{
   services.AddDbContext<MyDbContext>(options => options.UseSqlServer(configuration.GetConnectionString("string"));

}

```
Also your making a AuditTable that will be used to store all the aduits right? Is this including the high level ones? Would the json data just go into the Events?

**Comment by thepirat000** [comment]
`Setting.SqlConnectionString` was just an example to connect with a connection string.

Yes, the table will include all the events. 

Nothing stops you to try and see by yourself.

Audit.Core.Configuration.Setup() should be called on your app startup.

**Comment by chobo2** [comment]
I
o


> `Setting.SqlConnectionString` was just an example to connect with a connection string.
> 
> Yes, the table will include all the events.
> 
> Nothing stops you to try and see by yourself.
> 
> Audit.Core.Configuration.Setup() should be called on your app startup.

Okay I think I got it to work in my project. It seems to be logging though I am not sure if it doing the delete (or maybe how my code is setup maybe it does not work).


Some tables that are for drafts I clear the entire record and then reinsert the new record, I only see the insert not the clearing of the old record but maybe I missed it.

Also I need to attach to all high and low level events a userId, I cannot set that in the startup, am I able to call something per request and add this information in?

**Comment by chobo2** [comment]
I think the delete might be fine

still trying to see how I could dynamically add the userId to each CRUD request

**Comment by thepirat000** [comment]
it depends, how do you get the current userId?

**Comment by chobo2** [comment]
> it depends, how do you get the current userId?

Right now it would be through a claim via the http context on each request.
